File: backend/app/schemas/progress.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field

RATING_MIN = 0
RATING_MAX = 100


class ProgressCreate(BaseModel):
    month: date
    skill_rating: int = Field(default=0, ge=RATING_MIN, le=RATING_MAX)
    performance_rating: int = Field(default=0, ge=RATING_MIN, le=RATING_MAX)
    discipline_rating: int = Field(default=0, ge=RATING_MIN, le=RATING_MAX)
    remarks: str | None = None


class ProgressUpdate(BaseModel):
    skill_rating: int | None = Field(default=None, ge=RATING_MIN, le=RATING_MAX)
    performance_rating: int | None = Field(default=None, ge=RATING_MIN, le=RATING_MAX)
    discipline_rating: int | None = Field(default=None, ge=RATING_MIN, le=RATING_MAX)
    remarks: str | None = None


class ProgressOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    student_id: int
    month: date
    skill_rating: int
    performance_rating: int
    discipline_rating: int
    overall_score: float
    attendance_percentage: float | None = None
    remarks: str | None = None


class ProgressWithStudent(ProgressOut):
    student_name: str = ""


class ProgressList(BaseModel):
    items: list[ProgressWithStudent]
    total: int

File: mobile/lib/models/progress.dart
class MonthlyProgress {
  final int id;
  final int studentId;
  final String studentName;
  final String month;
  final int skillRating;
  final int performanceRating;
  final int disciplineRating;
  final double overallScore;
  final double? attendancePercentage;
  final String? remarks;

  const MonthlyProgress({
    required this.id,
    required this.studentId,
    this.studentName = '',
    required this.month,
    required this.skillRating,
    required this.performanceRating,
    required this.disciplineRating,
    required this.overallScore,
    this.attendancePercentage,
    this.remarks,
  });

  factory MonthlyProgress.fromJson(Map<String, dynamic> json) => MonthlyProgress(
        id: json['id'] as int,
        studentId: json['student_id'] as int,
        studentName: json['student_name'] as String? ?? '',
        month: json['month'] as String,
        skillRating: json['skill_rating'] as int? ?? 0,
        performanceRating: json['performance_rating'] as int? ?? 0,
        disciplineRating: json['discipline_rating'] as int? ?? 0,
        overallScore: (json['overall_score'] as num?)?.toDouble() ?? 0,
        attendancePercentage: (json['attendance_percentage'] as num?)?.toDouble(),
        remarks: json['remarks'] as String?,
      );
}

class ProgressListResponse {
  final List<MonthlyProgress> items;
  final int total;

  const ProgressListResponse({required this.items, required this.total});

  factory ProgressListResponse.fromJson(Map<String, dynamic> json) =>
      ProgressListResponse(
        items: (json['items'] as List)
            .map((e) => MonthlyProgress.fromJson(e as Map<String, dynamic>))
            .toList(),
        total: json['total'] as int? ?? 0,
      );
}